enhancements/openstack-baremetal-compute.md

```diff
@@ -0,0 +1,141 @@
+---
+title: openstack-baremetal-compute
+authors:
+  - "@pierreprinetti"
+reviewers:
+  - "@openshift/team-openstack-cloud-provider"
+approvers:
+  - TBD
+creation-date: 2019-12-10
+last-updated: 2019-12-10
+status: provisional
+
+---
+
+# Openstack Baremetal Compute
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+OpenStack's [Nova][openstack-nova] service can be configured to provision
+baremetal machines as OpenShift Compute Nodes.
+
+This enhancement will allow [Ironic][openstack-ironic]-provisioned machines to
+be used as OpenShift Compute Nodes, when they are made available as Nova
+flavors.
+
+## Motivation
+
+In heavy-loaded clusters, or when the workload is better handled with specially
+purposed hardware, the functional requirement of controlling where the work is
+computed might outweight the added value of virtualisation.
+
+In OpenStack clusters where Nova transparently provisions baremetal machines,
+this enhancement will ensure that OpenShift installations can use them as
+Compute Nodes.
+
+### Goals
+
+* Enable running OpenShift on a mix of virtual and baremetal machines.
+* Baremetal machines can be added and removed from a running OpenShift cluster.
+
+### Non-Goals
+
+An OpenStack setting where baremetal machines can't be attached to a virtual
+subnet is not supported in this enhancement.
```

Comment:
> I'm not sure this is actually possible. AFAIK, properly configured Ironic in the overcloud will provide connectivity between VMs in a subnet and the baremetal machines it boots up, but I'm not sure it will actually place them in the same subnet.

@dtantsur Can you please help here? Is "VMs and BMs in the same subnet" an impossible (or unlikely-to-be-fulfilled) requirement?

Reply:
Looks like it is actually possible (using the `neutron` network interface rather than the `flat` one):

https://docs.openstack.org/ironic/pike/admin/multitenancy.html

However, I think we may have to support both (or choose based on the relative use in the field).

Reply:
I would go with flat/vlan networks first, just because once you start using the neutron option you start to deal with having to configure real physical switches, which is possible with some of the ML2 drivers like network-generic-switch (https://docs.openstack.org/networking-generic-switch/latest/) or networking-ansible (https://networking-ansible.readthedocs.io/en/latest/), but once you do that, you start to have to deal with "compatability" of the switches. In other words, you make an assumption that the openstack cloud you are deploying to uses switches that work with those two mehanisms. It starts to get a get uncertain you can support all openstack clouds out there.

Reply:
Do all the nodes vm and bm in a cluster need to listen to the same ip broadcast (subnet broadcast)? If not it could be possible to link two subnet via a neutron router, if all vms and bms get a default gateway, it could be possible to place them in two different subnets. Then maybe you could connect a flat bm subnet to a vm tenant network.

Reply:
@adduarte I am 99% sure they don't need to live on the same broadcast domain as long as multicast DNS is not used. Note, for some functionality multicast dns was used, but regardless the ignition configuration should be supplied via configuration drive, which is the key to getting the cluster up and passing the configuration along. Honestly, and I stressed this in discussion with @tomassedovic ealier today, that the focus really just needs to be on providing what network to be operating on as supplied by the operator/installer/admin. The installer shouldn't need to know about nor care about implementation details of the cloud or baremetal network connection architecture, it should just be a network passed along and the machines are attached to that network.

I hope I'm making sense, since I've been banging my head against CI jobs all day.

Reply:
So it sounds like the openshift compute nodes could theoretically be in different subnets. 
If that is the case, could we not support BM openshift compute nodes along with vm openshift compute nodes by simply supporting the case where openshift compute nodes are in different subnets? 
I mean to say, that if you can support the nodes in separate subnets connected by a router, then you can first start by supporting BM nodes on an ironic subnet (flat, vlan, whatever is possible). Then once that is done and working it should be easy to simply support vms and bm on the same deployment by simply asking that the user put the BMs in one subnet, the VMs on another, and connect those subnets with a router.  The installer wouldn't care or need to know. It just needs the network info on a per node basis as its stated somewhere else here.

Reply:
@adduarte I've pushed an update yesterday that should address this. That is exactly what the new proposal outlines (barring errors in understanding or communication on my part).

The current proposal should not care what networking interface Ironic uses and when spinning up the baremetal workers, we will pass in the requisite flavor and network.